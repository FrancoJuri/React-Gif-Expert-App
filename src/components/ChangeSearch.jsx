import { useState } from 'react';
import PropTypes from 'prop-types';

const ChangeSearch = ({ onChangeSearch }) => {

    const [inputValue, setInputValue] = useState('');

    const onInputChange = ({ target }) => {
        setInputValue(target.value)
    }

    const onSubmit = (e) => {
        e.preventDefault();

        const newInputValue = inputValue.trim();
        
        if(newInputValue.length === 0){
            return;
        }

        onChangeSearch(newInputValue);
        setInputValue('');
    }

    return (
        <form className='search-form' onSubmit={onSubmit}>
            <input 
                type='text' 
                placeholder='Search for Gifs'  
                value={inputValue}
                onChange={onInputChange}
            />
            <button className='material-symbols-outlined pointer' type='submit'>
                search
            </button>
        </form>
    )
}

ChangeSearch.propTypes = {
    onChangeSearch: PropTypes.func.isRequired,
}

export default ChangeSearch;