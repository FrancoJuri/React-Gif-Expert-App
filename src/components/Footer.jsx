
const Footer = () => {
    return (
        <footer>
            <h3 className='text-center'>Made with ‚ù§ by Franco Juri</h3>
        </footer>
    )
}

export default Footer;