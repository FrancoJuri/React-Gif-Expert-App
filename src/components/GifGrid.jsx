import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifItem, Pagination, Footer } from './';
import PropTypes from 'prop-types';

const GifGrid = ({ search }) => {

    const { gifs, isLoading, page, setPage, pagination } = useFetchGifs(search);

    const onPreviousPage = () => {
        setPage(page - 1);
    }

    const onNextPage = () => {
        setPage(page + 1);
    }

    return (
        <>
            
            <h2 className='search-title text-center' data-testid='searchTitle'>{search}</h2>

            {
                isLoading && <h2 className='text-center'>Loading...</h2>
            }

            <div className='card-grid'>
                {
                    gifs.map((gif) => (
                        <GifItem 
                            key={ gif.id } 
                            {...gif}
                        />
                    ))
                }

                {
                    (!gifs.length && !isLoading)
                    &&
                    <div className='no-gifs'>
                        <h2 className='text-center'>No Gifs Found</h2>
                    </div>
                }
            </div>
            

            {
                (!isLoading && gifs.length >= 1)
                &&
                <>
                    <Pagination 
                        onPreviousPage={onPreviousPage}
                        onNextPage={onNextPage}
                        page={page}
                        pagination={pagination}
                    />
                    <Footer />
                </>
            }

        </>
    )
}

GifGrid.propTypes = {
    search: PropTypes.string.isRequired,
}

export default GifGrid;