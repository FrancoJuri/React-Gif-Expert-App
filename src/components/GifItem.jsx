import PropTypes from 'prop-types';

const GifItem = ({ title, url }) => {
    return (
        <div className='card'>
            <a href={url} target='_blank'>
                <img src={url} alt={title} />
            </a>
            <p>{ title }</p>
        </div>
    )
}

GifItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}

export default GifItem;