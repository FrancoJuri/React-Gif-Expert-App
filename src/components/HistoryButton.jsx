import { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

const HistoryButton = ({ setIsModalActive }) => {

  const [isFooterVisible, setIsFooterVisible] = useState(false);

  const checkFooter = () => {
    document.addEventListener('scroll', () => {

      if(!document.querySelector('footer')) return;

      const observer = new IntersectionObserver((entries) => {
        const [entry] = entries;
        setIsFooterVisible(entry.isIntersecting);
      })

      observer.observe(document.querySelector('footer'));
    })
  }

  useEffect(() => {

    if(!window.matchMedia('(max-width: 680px)').matches){
      return;
    }

    checkFooter();

    return () => {
      document.removeEventListener('scroll', checkFooter);
    }
    
  }, []) 

  let component;

  if(isFooterVisible){
    component = null;
  } else{
    component = <button className='history-button pointer' onClick={() => {
      setIsModalActive(true);
    }}>
        <span className='material-symbols-outlined'>history</span>
        <span className='history-button-text'>History</span>
    </button>;
  }

  return (component)
}

HistoryButton.propTypes = {
  setIsModalActive: PropTypes.func.isRequired,
}

export default HistoryButton;