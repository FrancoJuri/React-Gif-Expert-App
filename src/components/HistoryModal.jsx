import PropTypes from 'prop-types';

const HistoryModal = ({ history, onChangeSearch, onDeleteHistoryItem, onDeleteAllHistory, setIsModalActive }) => {
    return (
        <div className='modal'>
            <div className='modal-container'>

                <div className="modal-header">
                    <h3 className='modal-title'> 
                        <span className="material-symbols-outlined">history</span> 
                        History 
                    </h3>
                </div>

                <hr />

                <div className='modal-body'>
                    <ul className='history-list styled-scrollbar'>
                        <div>
                            {
                                history.map(item => (
                                    <li className='history-item pointer' key={item} onClick={(e) => {
                                        e.stopPropagation();
                                        onChangeSearch(item);
                                        setIsModalActive(false);
                                    }}>
                                        {item}
                                        <span className='material-symbols-outlined pointer' onClick={(e) => {
                                            e.stopPropagation();
                                            onDeleteHistoryItem(item);
                                        }}>
                                            delete
                                        </span>
                                    </li>
                                ))
                            }
                        </div>

                        {
                            (history.length >= 2)
                            &&
                            <button className="remove-history pointer" onClick={onDeleteAllHistory}>
                                Delete All
                            </button>
                        }

                        {
                            (!history.length)
                            &&
                            <p className='text-center'>Your history is empty</p>
                        }
                    </ul>
                </div>

                <hr />

                <div className="modal-footer">
                    <button className='modal-close pointer' onClick={() => {
                        setIsModalActive(false);
                    }}>Close</button>
                </div>

            </div>
        </div>
    )
}

HistoryModal.propTypes = {
    history: PropTypes.array.isRequired,
    onChangeSearch: PropTypes.func.isRequired,
    onDeleteHistoryItem: PropTypes.func.isRequired,
    onDeleteAllHistory: PropTypes.func.isRequired,
    setIsModalActive: PropTypes.func.isRequired,
}

export default HistoryModal;