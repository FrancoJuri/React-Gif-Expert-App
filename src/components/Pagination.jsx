import PropTypes from 'prop-types';

const Pagination = ({ onPreviousPage, onNextPage, page, pagination }) => {

    const totalCount = pagination.current.total_count;

    let totalPages;
    if(totalCount % 10 === 0){
        totalPages = (totalCount / 10) - 1;
    } else{
        totalPages = Math.floor(totalCount / 10)
    }

    return (
        <div className='pagination'>

            {
                (page !== 0)
                &&
                <button className='pointer' onClick={onPreviousPage}>
                    <span className='icon material-symbols-outlined'>
                        navigate_before
                    </span>
                </button>
            }

            <p>{page + 1} of {totalPages + 1}</p>

            {
                (page !== totalPages)
                &&
                <button className='pointer' onClick={onNextPage}>
                    <span className='icon material-symbols-outlined'>
                        navigate_next
                    </span>
                </button>
            }

        </div>
    )
}

Pagination.propTypes = {
    onPreviousPage: PropTypes.func.isRequired,
    onNextPage: PropTypes.func.isRequired,
    page: PropTypes.number.isRequired,
    pagination: PropTypes.object.isRequired,
}

export default Pagination;