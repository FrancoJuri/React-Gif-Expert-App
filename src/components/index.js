
export { default as ChangeSearch } from './ChangeSearch';
export { default as GifGrid } from './GifGrid';
export { default as GifItem } from './GifItem';
export { default as Pagination } from './Pagination';
export { default as Footer } from './Footer';
export { default as HistoryButton } from './HistoryButton';
export { default as HistoryModal } from './HistoryModal';