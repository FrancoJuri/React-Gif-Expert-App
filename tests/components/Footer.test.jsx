import { render } from '@testing-library/react'
import { Footer } from '../../src/components';


describe('Footer testing', () => {

    test('Should match snapshot', () => {

        const { container } = render(<Footer />);

        expect(container).toMatchSnapshot();
    })

})