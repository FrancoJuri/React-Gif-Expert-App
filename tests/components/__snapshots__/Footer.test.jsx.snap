// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer testing Should match snapshot 1`] = `
<div>
  <footer>
    <h3
      class="text-center"
    >
      Made with ‚ù§ by Franco Juri
    </h3>
  </footer>
</div>
`;
